### PR TITLE
feat: lr scheduler

### PR DESCRIPTION
**Pull request by Chamberlain0w0**
2025 年训练营项目选题“学习率调度器实现”中， @littleotherut 完成了学习率调度器模块的基本实现（PR #113 ）。这个 PR 是在他实现的基础上进一步修改接口、规范相关，使之符合我们项目实际应用的需求。

设计文档可以参考：https://gxtctab8no8.feishu.cn/wiki/Bd6Pw1BeeiQ7QfktiT8cbSiTnFb?from=from_copylink。

核心修改：
1. **接口上的修改**：采用 megatron 风格的学习率调度相关参数，`main.cc` 里面加了相应的 gflags。
```cpp
DEFINE_double(learning_rate, ..., "Peak learning rate.");
DEFINE_double(min_lr, 0.0, "Minimum learning rate.");
DEFINE_string(lr_decay_style, "constant", "LR decay style: none|constant|linear|cosine|inverse-square-root");
DEFINE_int64(lr_warmup_iters, 0, "Number of linear warmup iterations.");
DEFINE_double(lr_warmup_init, 0.0, "Initial learning rate at the start of warmup.");
DEFINE_int64(lr_decay_iters, 0, "Number of iterations to decay LR over (0 = num_iteration).");
```
获得上述参数后，构造相应的 `TrainingLRSchedulerConfig` 结构体，然后同优化器一并传给 `CreateLRScheduler()` 构造得到对应的学习率调度器。

2. **设计上的修改**：
    a. 新增 `LRScheduler` 基类以及各种调度策略对应的派生类，与 torch 实现以及使用方式均对齐。训练循环中在 `optimizer.step()` 之后再调用 `scheduler.step()` 即完成学习率更新。
    b. `LRScheduler` 需要与 `Optimizer` 交互来获取、更新、同步学习率，所以给 `Optimizer` 基类也加了对应的 setter 和 getter。
    c. `LRScheduler::State()` 的部分仍是一个较 naive 的实现，后续等 ckpt 机制完成以后再进一步修改。

### CODE REVIEW

File: infini_train/src/lr_scheduler.cc [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+float ConstantLR::GetClosedFormLR() const { return last_step_ < total_iters_ ? base_lr_ * factor_ : base_lr_; }
+
+float ConstantLR::GetChainedFormLR() const {
+    const float lr = optimizer_->learning_rate();
+    if (last_step_ == 0) {
+        return lr * factor_;
+    } else if (last_step_ < total_iters_) {
+        return lr;
+    } else if (last_step_ == total_iters_) {
+        return lr / factor_;
```
Comment (JYMiracle305) [inline infini_train/src/lr_scheduler.cc:173]: 对factor_的检查是 >= 0，这里有除以0的风险

File: infini_train/include/lr_scheduler.h [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+    void Step() override;
+    void InitialStep() override;
+
+    StateDict State() const override;
+    void LoadState(const StateDict &state) override;
+
+protected:
+    // FIXME: SequentialLR should not have a closed-form LR, but we need to implement this pure virtual function.
+    float GetClosedFormLR() const override { return base_lr_; }
```
Comment (JYMiracle305) [inline infini_train/include/lr_scheduler.h]: 这里直接LOG(FATAL)更合适，也防止误调用

File: infini_train/include/lr_scheduler.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    void Step() override;
+    void InitialStep() override;
+
+    StateDict State() const override;
+    void LoadState(const StateDict &state) override;
+
+protected:
+    // FIXME: ChainedScheduler should not have a closed-form LR, but we need to implement this pure virtual function.
+    float GetClosedFormLR() const override { return base_lr_; }
```
Comment (JYMiracle305) [inline infini_train/include/lr_scheduler.h]: 同上

File: infini_train/src/lr_scheduler.cc (hunk 1 above)
Comment (Chamberlain0w0) [reply infini_train/src/lr_scheduler.cc:173]: image

这块 [torch ](https://github.com/pytorch/pytorch/blob/ef303cde7da336bcf3c43bf8f8211ac1effbf606/torch/optim/lr_scheduler.py#L813)的对 factor 的约束还真是 [0,1] ，而且我也没看到任何避免除 0 的操作，当时想着跟 torch 对齐；不过从数值正确性上看确实得避免除 0，这块没必要强行对齐。我改一下 check

File: example/gpt2/main.cc [hunk 4]
```diff
@@ -383,6 +389,7 @@ void Train(const nn::parallel::Rank &rank) {
             .rank = rank,
             .model = *model,
             .optimizer = *optimizer,
+            .lr_scheduler = scheduler.get(),
```
Comment (JYMiracle305) [inline example/gpt2/main.cc:407]: 这里没有传入开关 save_lr_scheduler_state，内部使用了默认值 true，确认一下需不需要传入做成可控的

File: infini_train/src/checkpoint/checkpoint.cc [hunk 5]
```diff
@@ -80,12 +200,17 @@ void Checkpoint::Save(const std::filesystem::path &checkpoint_dir, const nn::Mod
         }
     }
 
+    if (save_lr_scheduler_state && lr_scheduler != nullptr) {
```
Comment (JYMiracle305) [inline infini_train/src/checkpoint/checkpoint.cc]: 建议 lr_scheduler 指针的判断也放在内部CHECK，如果打开 save_lr_scheduler_state，指针为空直接报错可以感知当前不能保存学习率调度

File: infini_train/src/checkpoint/checkpoint.cc [hunk 6]
```diff
@@ -103,6 +228,18 @@ void Checkpoint::Load(const std::filesystem::path &checkpoint_dir, nn::Module &m
     }
 
     state = LoadTrainerState(checkpoint_dir / "trainer_state.json");
+
+    if (load_lr_scheduler_state && lr_scheduler != nullptr) {
```
Comment (JYMiracle305) [inline infini_train/src/checkpoint/checkpoint.cc]: 同上

File: example/gpt2/main.cc
Comment (JYMiracle305) [inline example/gpt2/main.cc:366]: 确认是否需要传入 load_lr_scheduler_state

File: example/gpt2/main.cc [hunk 7]
```diff
@@ -367,7 +373,7 @@ void Train(const nn::parallel::Rank &rank) {
             .save_dir = save_dir,
             .global_step = global_step,
             .consumed_batches = consumed_batches,
-            .last_lr = FLAGS_learning_rate,
+            .last_lr = optimizer->learning_rate(),
```
Comment (JYMiracle305) [inline example/gpt2/main.cc]: 可以把这个字段删掉了，这个字段一直没有使用。之前 lr 不会动态改变；现在依赖 lr scheduler 恢复 lr。

File: example/gpt2/main.cc (hunk 4 above)
Comment (Chamberlain0w0) [reply example/gpt2/main.cc:407]: 需要传，我改一下。

File: infini_train/src/checkpoint/checkpoint.cc (hunk 5 above)
Comment (Chamberlain0w0) [reply infini_train/src/checkpoint/checkpoint.cc]: ok 我改一下

File: infini_train/src/checkpoint/checkpoint.cc (hunk 6 above)
Comment (Chamberlain0w0) [reply infini_train/src/checkpoint/checkpoint.cc]: ok 我改一下

File: example/gpt2/main.cc
Comment (Chamberlain0w0) [reply example/gpt2/main.cc:366]: 需要，我改一下

File: example/gpt2/main.cc (hunk 7 above)
Comment (Chamberlain0w0) [reply example/gpt2/main.cc]: ok，我把相关 FIXME 都删了

File: example/gpt2/main.cc (hunk 4 above)
Comment (JYMiracle305) [reply example/gpt2/main.cc:407]: .load_lr_scheduler_state = scheduler != nullptr   这里依赖 scheduler 判空赋值，再加上内部判断load_lr_scheduler_state 后再判断 scheduler 是否为空，实际上没有必要，可以不加 save_lr_scheduler_state 。我看了一下megatron，也没有保存 lr_scheduler 的开关
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/training/checkpointing.py#L1076

Comment (Chamberlain0w0) [reply example/gpt2/main.cc:407]: 确实略显重复，我删一下吧，统一按 lr_scheduler 做指针非空判断

**Review by JYMiracle305 — APPROVED** [review]

File: infini_train/include/checkpoint/checkpoint.h [hunk 8]
```diff
@@ -34,10 +32,10 @@ struct TrainerState {
 class Checkpoint {
 public:
     static void Save(const std::filesystem::path &checkpoint_dir, const nn::Module &model, const Optimizer *optimizer,
-                     const TrainerState &state, bool save_optimizer_state);
+                     const TrainerState &state, bool save_optimizer_state, const LRScheduler *lr_scheduler = nullptr);
```
Comment (kilinchange) [inline infini_train/include/checkpoint/checkpoint.h]: 建议不要在头文件声明中设置默认值，降低后续维护成本。若调用方允许为空，在调用层显式传入 nullptr，Load 同理。

File: infini_train/include/lr_scheduler.h [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    virtual void Step();
+    virtual void Step(int64_t epoch);
+    virtual void InitialStep();
+
+    float GetLR() const;
+    float BaseLR() const;
+    int64_t LastStep() const;
+
+    void ResetStep(int64_t step = -1);
+    virtual StateDict State() const;
```
Comment (kilinchange) [inline infini_train/include/lr_scheduler.h]: 命名为 StateDict？
下面 load 接口也是，LoadStateDict

File: infini_train/include/lr_scheduler.h [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    virtual void LoadState(const StateDict &state);
+
+    bool SharesOptimizerWith(const std::shared_ptr<Optimizer> &opt) const;
```
Comment (kilinchange) [inline infini_train/include/lr_scheduler.h]: 这里提供个 optimizer() getter 接口就行，由调用方直接比较 optimizer 指针，语义上更直观一些，torch 也是这么比较的：
https://github.com/pytorch/pytorch/blob/c2b2e5d06dd357bd5f43b9ab7f7b7e3e6b4f568f/torch/optim/lr_scheduler.py#L1529

File: infini_train/include/lr_scheduler.h [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    LRScheduler(const LRScheduler &) = delete;
+    LRScheduler &operator=(const LRScheduler &) = delete;
+
+    virtual void Step();
+    virtual void Step(int64_t epoch);
+    virtual void InitialStep();
+
+    float GetLR() const;
+    float BaseLR() const;
+    int64_t LastStep() const;
```
Comment (kilinchange) [inline infini_train/include/lr_scheduler.h]: getter 接口和变量名保持一致

File: infini_train/include/lr_scheduler.h [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+protected:
+    virtual float GetClosedFormLR() const = 0;
+    virtual float GetChainedFormLR() const;
+    void ApplyLR(float lr);
+
+    std::shared_ptr<Optimizer> optimizer_;
+    int64_t last_step_;
+    float recover_lr_;
+    float base_lr_;
```
Comment (kilinchange) [inline infini_train/include/lr_scheduler.h]: 简单类型赋初值，这个文件里其他变量也是。

File: infini_train/include/optimizer.h [hunk 13]
```diff
@@ -27,8 +27,19 @@ class Optimizer {
...
+
+    float initial_learning_rate() const;
+
+    void set_initial_learning_rate(float lr);
+
 protected:
     std::vector<std::shared_ptr<Tensor>> params_;
+    float learning_rate_ = 0.0f;
+    float initial_learning_rate_ = 0.0f;
+    bool initial_lr_set_ = false;
```
Comment (kilinchange) [inline infini_train/include/optimizer.h:44]: optimizer 里有必要存 initial_learning_rate_ 吗？目前的场景里初始学习率是由 scheduler 的 base_lr 维护和使用的，似乎 optimizer 中没有必要额外维护？

File: infini_train/src/lr_scheduler.cc [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+        auto sub_state = schedulers_[i]->State();
+        for (const auto &[key, value] : sub_state) { state["scheduler_" + std::to_string(i) + "." + key] = value; }
+    }
+    return state;
+}
+
+void SequentialLR::LoadState(const StateDict &state) {
+    last_step_ = std::get<int64_t>(state.at("last_step"));
+    recover_lr_ = std::get<float>(state.at("recover_lr"));
+    base_lr_ = std::get<float>(state.at("base_lr"));
```
Comment (kilinchange) [inline infini_train/src/lr_scheduler.cc]: 这些 state key 建议抽成常量，避免硬编码字符串分散在代码中。

File: infini_train/src/optimizer.cc [hunk 15]
```diff
@@ -9,16 +9,32 @@
...
+float Optimizer::learning_rate() const { return learning_rate_; }
+
+float Optimizer::initial_learning_rate() const {
+    CHECK(initial_lr_set_) << "Optimizer: initial_learning_rate not set. "
+                              "Use with an LRScheduler first.";
+    return initial_learning_rate_;
+}
+
+void Optimizer::set_initial_learning_rate(float lr) {
+    if (!initial_lr_set_) {
```
Comment (kilinchange) [inline infini_train/src/optimizer.cc]: 是不是应该 CHECK？如果已经 set 过了就不允许再次被 set，不应该静默跳过。

File: infini_train/include/lr_scheduler.h (hunk 11 above)
Comment (Chamberlain0w0) [reply infini_train/include/lr_scheduler.h]: ok，这块我改一下，这块是之前写文档的时候引用了 torch 里面 lr scheduler 的接口名称，实现上就沿用了

File: infini_train/include/checkpoint/checkpoint.h (hunk 8 above)
Comment (Chamberlain0w0) [reply infini_train/include/checkpoint/checkpoint.h]: 好的，我改一下

File: infini_train/include/lr_scheduler.h (hunk 9 above)
Comment (Chamberlain0w0) [reply infini_train/include/lr_scheduler.h]: 改了

File: infini_train/include/lr_scheduler.h (hunk 12 above)
Comment (Chamberlain0w0) [reply infini_train/include/lr_scheduler.h]: 改了

File: infini_train/include/optimizer.h (hunk 13 above)
Comment (Chamberlain0w0) [reply infini_train/include/optimizer.h:44]: 参考：PR #113：https://github.com/InfiniTensor/InfiniTrain/pull/113/changes#r2938263711

这部分和学员讨论了下决定与 torch 命名一致，避免歧义。这个是有用的，是存在 Optimizer 里的“原始学习率快照”。它表示第一次创建 scheduler 时，optimizer 还没被 scheduler 改动前的 lr。因为目前组合 scheduler（如 SequentialLR）的实现中，多个 scheduler 共享同一个 optmizer，第一个 scheduler 构造时把 optimizer 的原始 lr 固定为 initial lr（注意 `set_initial_learning_rate` 操作有个 flag 保证只做一次），后面每个 scheduler 构造时，都从这个固定的 initial_lr 的值复制出自己的 `base_lr_`，而不是从已经变化的 `optimizer.learning_rate_` 读。

File: infini_train/src/lr_scheduler.cc (hunk 14 above)
Comment (Chamberlain0w0) [reply infini_train/src/lr_scheduler.cc]: 好的，我改一下

File: infini_train/src/optimizer.cc (hunk 15 above)
Comment (Chamberlain0w0) [reply infini_train/src/optimizer.cc]: 参考上面的：

> 因为目前组合 scheduler（如 SequentialLR）的实现中，多个 scheduler 共享同一个 optmizer，第一个 scheduler 构造时把 optimizer 的原始 lr 固定为 initial lr（注意 set_initial_learning_rate 操作有个 flag 保证只做一次），后面每个 scheduler 构造时，都从这个固定的 initial_lr 的值复制出自己的 base_lr_，而不是从已经变化的 optimizer.learning_rate_ 读。

目前在组合 LRScheduler 的逻辑下就是会多次调用到，但是可以把判断提前。已经改成为在 `set_initial_learning_rate` 里做 CHECK，然后在 LRScheduler 构造函数里通过 flag 判断只 set 一次的逻辑。

File: infini_train/include/lr_scheduler.h (hunk 10 above)
Comment (Chamberlain0w0) [reply infini_train/include/lr_scheduler.h]: 改了

File: infini_train/include/optimizer.h (hunk 13 above)
Comment (Chamberlain0w0) [reply infini_train/include/optimizer.h:44]: 想了下，`optimizer.initial_learning_rate` 不需要单独保存在 optimizer checkpoint 里。它只是 scheduler 构造时用来确定 `base_lr_` 的；真正 resume 需要的 scheduler 基准 lr 已经作为 `base_lr` 存在 lr_scheduler.ckpt 里，当前 lr 也作为 `recover_lr` 保存并恢复了。resume 时恢复 scheduler 的流程是，先构造 scheduler 对象（此时确实会读 optimizer 的 `initial_learning_rate_`），但随后读 ckpt 参数，scheduler 自己保存的 `base_lr_` 会覆盖前者，所以 resume 场景下这个 initial lr 根本用不上。

File: infini_train/include/lr_scheduler.h (hunk 12 above)
Comment (kilinchange) [reply infini_train/include/lr_scheduler.h]: 这个文件里的其他类没改。

Comment (Chamberlain0w0) [reply infini_train/include/lr_scheduler.h]: 重新改了一下

**Comment by Chamberlain0w0** [comment]
老的 test cases，与 master 比较（在 3026 上截的，但是的确是在 3025 上跑的）：

截屏2026-07-02 22 37 47

截屏2026-07-02 22 39 29

**Comment by Chamberlain0w0** [comment]
与 torch 的正确性对齐：https://gxtctab8no8.feishu.cn/wiki/EsyzwGCrYiXj4XklTJ6cmwaunEh?from=from_copylink

参考： https://github.com/InfiniTensor/InfiniTrain-Test/pull/1

**Review by kilinchange — APPROVED** [review]
